### DISCUSSION TITLE
How do the devs do it?

### DISCUSSION THREAD

**Discussion by AZthemute** [Melledy/Grasscutter#1081] [Q&A]
Asking as a complete noob to this sort of server re-implementation stuff, I'm curious how everyone who works on this project knows what the game is asking from the server and how they reimplement the functionality.

...and maybe if I could do this also with Genshin or with another game.

**Comment by 4Benj** [comment]
A lot of guess work, packet sniffing, and reverse engineering (not to mention learning from other’s mistakes)

**Reply by AZthemute** [reply]
> A lot of guess work, packet sniffing

That's just about what I guessed, except guess work doesn't sound so fun.
> and reverse engineering

You mean reverse engineering the game client? We don't have the server so we can't reverse engineer that, obviously. In which case that brings up the question of how you would even start with reverse engineering the game client, but that doesn't really matter now.

**Reply by 4Benj** [reply]
> except guess work doesn't sound so fun.

Allow me to point you to the source of my insanity, [The mail PR](https://github.com/Grasscutters/Grasscutter/pull/115)...

**Reply by AZthemute** [reply]
> > except guess work doesn't sound so fun.
> 
> Allow me to point you to the source of my insanity, [The mail PR](https://github.com/Grasscutters/Grasscutter/pull/115)...

As someone who's never had to deal with this I can still feel your pain just by looking at the 18 comments...

Well, what if I do want to torture myself? What software/what do I need to know to start?
